File: src/boj/boj_2840.java
package boj;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Objects;
import java.util.StringTokenizer;

public class boj_2840 {
    public static void main(String[] args) throws IOException {
        final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        final int n = Integer.parseInt(st.nextToken());
        final int k = Integer.parseInt(st.nextToken());

        final ArrayDeque<Character> deque = new ArrayDeque<>();
        for (int i = 0; i < n; i ++) {
            deque.add('?');
        }

        for (int i = 0; i < k; i++) {
            st = new StringTokenizer(br.readLine());
            final int cnt = Integer.parseInt(st.nextToken());
            for (int j = 0; j < cnt; j++) {
                deque.addFirst(Objects.requireNonNull(deque.pollLast()));
            }
            final Character c = st.nextToken().charAt(0);
            final Character temp = deque.pollFirst();

            if ((temp != '?' && c != temp) || deque.contains(c)) {
                System.out.println("!");
                return;
            }
            deque.addFirst(c);
        }

        final StringBuilder answer = new StringBuilder();
        for (int i = 0; i < n; i++) {
            answer.append(deque.pollFirst());
        }
        System.out.println(answer);
    }
}

File: src/boj/boj_2840.kt
package boj

import java.util.*

fun main() = with(Scanner(System.`in`)) {
    val deque = ArrayDeque<Char>()
    val n = nextInt()
    for (i in 0 until n) {
        deque.add('?')
    }

    val k = nextInt()
    for (i in 0 until k) {
        val s = nextInt()
        for (j in 0 until s) {
            deque.addFirst(deque.pollLast())
        }
        val first = deque.pollFirst()
        val char = next().first()

        if ((first != '?' && first != char) || deque.contains(char)) {
            print('!')
            return@with
        }

        deque.addFirst(char)
    }

    for (c in deque) {
        print(c)
    }
}
